federation/pkg/federation-controller/service/dns.go

```diff
@@ -16,25 +16,230 @@ limitations under the License.
 	// Quinton: Pseudocode....
+	// See https://github.com/kubernetes/kubernetes/pull/25107#issuecomment-218026648
+	// For each service we need the following DNS names:
+	// mysvc.myns.myfed.svc.z1.r1.mydomain.com  (for zone z1 in region r1)
+	//         - an A record to IP address of specific shard in that zone (if that shard exists and has healthy endpoints)
+	//         - OR a CNAME record to the next level up, i.e. mysvc.myns.myfed.svc.r1.mydomain.com  (if a healthy shard does not exist in zone z1)
+	// mysvc.myns.myfed.svc.r1.federation
+	//         - a set of A records to IP addresses of all healthy shards in region r1, if one or more of these exist
+	//         - OR a CNAME record to the next level up, i.e. mysvc.myns.myfed.svc.mydomain.com (if no healthy shards exist in region r1)
+	// mysvc.myns.myfed.svc.federation
+	//         - a set of A records to IP addresses of all healthy shards in all regions, if one or more of these exist.
+	//         - no record (NXRECORD response) if no healthy shards exist in any regions)
+	//
+	// For each cached service, cachedService.lastState tracks the current known state of the service, while cachedService.appliedState contains
+	// the state of the service when we last successfully sync'd it's DNS records.
+	// So this time around we only need to patch that (add new records, remove deleted records, and update changed records.
+	//
+	// TODO: Some of these intermediate variables are not really necessary, but making sure that we have all of the required data for now.
+	serviceName := cachedService.lastState.Name
+	namespaceName := cachedService.lastState.Namespace
+	federationName := "myfederation" // TODO: quinton: Get this from the federation configuration
+	var (
+		zoneName, regionName, dnsZoneName string
+		err                               error
+	)
+	if zoneName, err = s.getClusterZoneName(clusterName); err != nil {
+		return err
+	}
+	if regionName, err = s.getClusterRegionName(clusterName); err != nil {
+		return err
+	}
+	if dnsZoneName, err = s.getFederationDNSZoneName(); err != nil {
+		return err
+	}
+	zoneEndpoints, regionEndpoints, globalEndpoints, err := s.getHealthyEndpoints(clusterName, cachedService)
+	if err != nil {
+		return err
+	}
+
+	commonPrefix := serviceName + "." + namespaceName + "." + federationName + ".svc"
+	// dnsNames is the path up the DNS search tree, starting at the leaf
+	dnsNames := []string{
+		commonPrefix + "." + zoneName + "." + regionName + "." + dnsZoneName, // zone level
+		commonPrefix + "." + regionName + "." + dnsZoneName,                  // region level, one up from zone level
+		commonPrefix + "." + dnsZoneName,                                     // global level, one up from region level
+		"", // nowhere to go up from global level
+	}
+
+	endpoints := [][]string{zoneEndpoints, regionEndpoints, globalEndpoints}
 
+	for i, endpoint := range endpoints {
+		err = s.ensureDnsRrsets(dnsZoneName, dnsNames[i], endpoint, dnsNames[i+1])
```

Comment:
See comment on ensureDnsRrsets:

```
/* ensureDnsRrsets ensures (idempotently, and with minimum mutations) that all of the DNS resourcerecordsets for dnsName are consistent with endpoints.
    if endpoints is nil or empty, a CNAME record to uplevelCname is ensured
 */
```

So all healthy endpoints are ensured to be there, and all unhealthy endpoints ensured to be absent.  At least that's the intention.  Make sense?